packages/apollo-engine-reporting/src/contextualizedStats.ts

```diff
@@ -0,0 +1,246 @@
+    queryLatencyStats.requestCount++;
+    if (trace.fullQueryCacheHit) {
+      queryLatencyStats.cacheLatencyCount.incrementDuration(trace.durationNs);
+      queryLatencyStats.cacheHits++;
+    } else {
+      queryLatencyStats.latencyCount.incrementDuration(trace.durationNs);
+    }
+
+    if (
+      !trace.fullQueryCacheHit &&
+      trace.cachePolicy &&
+      trace.cachePolicy.maxAgeNs
+    ) {
+      if (trace.cachePolicy.scope == Trace.CachePolicy.Scope.PRIVATE) {
+        queryLatencyStats.privateCacheTtlCount.incrementDuration(
+          trace.cachePolicy.maxAgeNs,
+        );
+      } else if (trace.cachePolicy.scope == Trace.CachePolicy.Scope.PUBLIC) {
+        queryLatencyStats.publicCacheTtlCount.incrementDuration(
+          trace.cachePolicy.maxAgeNs,
+        );
+      }
+    }
+
+    if (trace.persistedQueryHit) {
+      queryLatencyStats.persistedQueryHits++;
+    } else if (trace.persistedQueryRegister) {
+      queryLatencyStats.persistedQueryMisses++;
+    }
+
+    if (trace.forbiddenOperation) {
+      queryLatencyStats.forbiddenOperationCount++;
+    } else if (trace.registeredOperation) {
+      queryLatencyStats.registeredOperationCount++;
+    }
+
+    let hasError = false;
+    const typeStats = this.perTypeStat;
+    const rootPathErrorStats = queryLatencyStats.rootErrorStats;
+
+    function traceNodeStats(node: Trace.INode, path: ReadonlyArray<string>): boolean {
+      // Generate error stats and error path information
+      if (node.error && node.error.length > 0) {
+        hasError = true;
+
+        let currPathErrorStats = rootPathErrorStats;
+
+        for (const subPathEntry of path.entries()) {
+          // Using entries instead values since Node 8
+          // doesn't support Array.prototype.values
+          const subPath = subPathEntry[1];
+          let children =
+            currPathErrorStats.children ||
+            (currPathErrorStats.children = Object.create(null));
+          currPathErrorStats =
+            children[subPath] || (children[subPath] = Object.create(null));
+        }
+
+        currPathErrorStats.requestsWithErrorsCount =
+          (currPathErrorStats.requestsWithErrorsCount || 0) + 1;
+        currPathErrorStats.errorsCount =
+          (currPathErrorStats.errorsCount || 0) + node.error.length;
+      }
+
+      if (
+        node.parentType != null &&
+        node.originalFieldName != null &&
+        node.type != null &&
+        node.endTime != null &&
+        node.startTime != null
+      ) {
+        let typeStat = typeStats[node.parentType];
+        if (!typeStat) {
+          typeStat = typeStats[node.parentType] = new TypeStat();
+        }
+
+        let fieldStat = typeStat.perFieldStat[node.originalFieldName];
+        if (!fieldStat) {
+          fieldStat = typeStat.perFieldStat[
+            node.originalFieldName
+          ] = new FieldStat(node.type);
+        }
+
+        // We only create the object in the above line so we can know they aren't null
```

Comment:
not sure what this means... but note that I think TS is smart enough to narrow the type of `fieldStat` to be defined here when you use the `if (!x) { x = }` pattern. (You can see this if you explicitly declare `let fieldStat: FieldStat | undefined =`)